### PR TITLE
Customize queue behavior for pipeline waiting

### PR DESCRIPTION
**Pull request by jcusa**
Fixing the issue https://github.com/spinnaker/spinnaker/issues/805

Allowing queuing pipelines to stay in queue.

### CODE REVIEW

**Comment by tomaslin** [comment]
tests failed

**Comment by jcusa** [comment]
yeah. looking into it.

**Comment by jcusa** [comment]
Fixed the test.

**Comment by jcusa** [comment]
@tomaslin  I have fixed the test. Also, there's a related pull request here: https://github.com/spinnaker/spinnaker/pull/811

File: orca-core/src/test/groovy/com/netflix/spinnaker/orca/batch/pipeline/PipelineStarterListenerSpec.groovy [hunk 1]
```diff
@@ -63,6 +63,7 @@ class PipelineStarterListenerSpec extends Specification {
 
   def "should mark waiting pipelines that are not started as canceled"() {
     given:
+    listener.optimizeWaitingQueue = true
```
Comment (tomaslin) [inline orca-core/src/test/groovy/com/netflix/spinnaker/orca/batch/pipeline/PipelineStarterListenerSpec.groovy]: There should be a test here for the two behaviors

File: orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/PipelineStarterListener.groovy [hunk 2]
```diff
@@ -76,11 +81,13 @@ class PipelineStarterListener implements JobExecutionListener {
             def queuedExecution = executionRepository.retrievePipeline(id)
             log.info("starting pipeline ${nextPipelineId} due to ${execution.id} ending")
             pipelineStarter.startExecution(queuedExecution)
-          } else {
+            startTracker.removeFromQueue(execution.pipelineConfigId, id)
```
Comment (tomaslin) [inline orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/PipelineStarterListener.groovy]: not too keen on this being a global flag. Feels like you want to also be able to override this on a per pipeline config level.

Comment (jcusa) [inline orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/PipelineStarterListener.groovy]: @tomaslin yeah, thanks, that makes sense.  you are suggesting to have a UI checkbox in the individual pipeline configuration level for this, correct?

**Comment by jcusa** [comment]
@tomaslin yeah. thanks. that makes sense. you are saying, have a UI check box in the configuration of the pipeline

**Comment by jcusa** [comment]
Will update current PR making the flag at the individual pipeline level. Also, will have a deck PR where a checkbox will be available if 'Disable concurrent pipeline' is checked. 

sample screen shot: 
this is when 'Disable concurrent pipeline is checked' the new flag will show up for you to check 
newflag

this is when 'Disable concurrent pipeline is unchecked' 
newflag2

**Comment by tomaslin** [comment]
I would make that second checkbox only show up when the first one is selected. Visually, it feels like it should be indented so that the relationship to the parent checkbox is obvious ( might be optional if we're hiding and showing ).

**Comment by jcusa** [comment]
thanks tomas, i added screen shots above.

**Comment by tomaslin** [comment]
yeah, comments was related to screenshot. Also, make sure there is a help field explaining what it does

**Comment by jcusa** [comment]
yes. Will hide the check box if the first box is not checked. and, thanks for mentioning. will add help text too.

**Comment by jcusa** [comment]
related UI changes is here:
https://github.com/spinnaker/deck/pull/2119

**Comment by jcusa** [comment]
I have merged in the new changes. No longer using global config. The flag is at the pipeline level now.

File: orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/persistence/jedis/JedisExecutionRepository.groovy [hunk 3]
```diff
@@ -345,7 +345,8 @@ class JedisExecutionRepository implements ExecutionRepository {
       executingInstance: execution.executingInstance,
       status           : execution.status?.name(),
       authentication   : mapper.writeValueAsString(execution.authentication),
-      paused           : mapper.writeValueAsString(execution.paused)
+      paused           : mapper.writeValueAsString(execution.paused),
+      limitWaiting     : String.valueOf(execution.limitWaiting)
```
Comment (tomaslin) [inline orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/persistence/jedis/JedisExecutionRepository.groovy]: the variable name here is not very clear. can we change it to something more descriptive like keepWaitingPipelines

File: orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/PipelineStarter.groovy [hunk 4]
```diff
@@ -54,6 +54,7 @@ class PipelineStarter extends ExecutionStarter<Pipeline> {
       .withAppConfig((Map<String, Serializable>) config.appConfig)
       .withParallel(config.parallel as Boolean)
       .withLimitConcurrent(config.limitConcurrent as Boolean)
+      .withLimitWaiting(config.limitWaiting == null ? true : config.limitWaiting as Boolean)
```
Comment (tomaslin) [inline orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/PipelineStarter.groovy]: how about ```config.limitWaiting ?: true```

**Comment by jcusa** [comment]
thank you. changed variable name per suggestion to "keepWaitingPipelines". with that, default is now false  (i.e.  by default pipelines will not wait in queue)

File: orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/PipelineStarter.groovy (hunk 4 above)
Comment (ajordens) [inline orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/PipelineStarter.groovy]: That probably won't handle the case where `config.limitWaiting == false`.

Comment (jcusa) [inline orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/PipelineStarter.groovy]: agree. 
in the latest code, i ended up not needing the null check due to changing of the variable name/meaning per tomas' suggestion. so i think we are good here.

**Comment by anotherchrisberry** [comment]
@tomaslin good here?

**Comment by tomaslin** [comment]
LGTM

**Pull request merged**